pytest-server-fixtures/pytest_server_fixtures/s3.py

```diff
@@ -0,0 +1,94 @@
+# coding: utf-8
+"""
+Pytest fixtures to launch a minio S3 server and get a bucket for it.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import uuid
+
+import boto3
+import botocore.client
+import pytest
+from future.utils import text_type
+from pytest_fixture_config import requires_config
+
+from . import CONFIG
+from .base import TestServer
+
+
+def _s3_server(request):
+    server = MinioServer()
+    server.start()
+    request.addfinalizer(server.teardown)
+    return server
+
+
+@requires_config(CONFIG, ['minio_bin'])
+@pytest.fixture(scope="session")
+def s3_server(request):
+    return _s3_server(request)
+
+
+# Minio is a little too slow to start for each function call
+# Start it once per session and get a new bucket for each function instead.
+@pytest.fixture(scope="function")
+def s3_bucket(s3_server):  # pylint: disable=redefined-outer-name
+    client = s3_server.get_s3_client()
+    bucket_name = text_type(uuid.uuid4())
+    client.create_bucket(Bucket=bucket_name)
+    return client, bucket_name
+
+
+class MinioServer(TestServer):
+    random_port = True
+    aws_access_key_id = "MINIO_TEST_ACCESS"
+    aws_secret_access_key = "MINIO_TEST_SECRET"
+
+    def __init__(self, workspace=None, delete=None, preserve_sys_path=False, **kwargs):
+        env = kwargs.get('env', {})
+        env.update({"MINIO_ACCESS_KEY": self.aws_access_key_id, "MINIO_SECRET_KEY": self.aws_secret_access_key})
+        kwargs['env'] = env
+        super(MinioServer, self).__init__(workspace, delete, preserve_sys_path, **kwargs)
+
+    def get_s3_client(self):
+        # Region name and signature are to satisfy minio
+        s3 = boto3.resource(
+            's3',
+            endpoint_url=self.boto_endpoint_url,
+            aws_access_key_id=self.aws_access_key_id,
+            aws_secret_access_key=self.aws_secret_access_key,
+            region_name='us-east-1',
+            config=botocore.client.Config(signature_version='s3v4'),
+        )
+        return s3
+
+    @property
+    def datadir(self):
+        return self.workspace / 'minio-db'
+
+    @property
+    def boto_endpoint_url(self):
+        return "http://localhost:{port}".format(port=self.port)
+
+    def pre_setup(self):
+        self.datadir.mkdir()  # pylint: disable=no-value-for-parameter
+
+    @property
+    def run_cmd(self):
+        cmdargs = [
+            CONFIG.minio_bin,
+            "server",
+            "--address",
+            ":{}".format(self.port),
+            text_type(self.datadir),
+        ]
+        return cmdargs
+
+    def check_server_up(self):
+        return True  # TODO do something smarter here
```

Comment:
I don't know exactly what to do here. So far this has been working for me and has been stable. I'll see if there's a lightweight ping-style command I can ru via the client.